beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+// Since DataColumnStorage only writes data columns that have undergone full verification, the return
+// value is always a VerifiedRODataColumn.
+func (dcs *DataColumnStorage) Get(root [fieldparams.RootLength]byte, indices []uint64) ([]blocks.VerifiedRODataColumn, error) {
+	dcs.pruneMu.RLock()
+	defer dcs.pruneMu.RUnlock()
+
+	fileMu, _ := dcs.fileMutex(root)
+	fileMu.RLock()
+	defer fileMu.RUnlock()
+
+	startTime := time.Now()
+
+	// Build all indices if none are provided.
+	if indices == nil {
+		indices = make([]uint64, mandatoryNumberOfColumns)
+		for i := range indices {
+			indices[i] = uint64(i)
+		}
+	}
+
+	// Preventive check of indices.
+	for _, index := range indices {
+		if index >= mandatoryNumberOfColumns {
+			return nil, ErrDataColumnIndexTooLarge
+		}
+	}
+
+	summary, ok := dcs.cache.get(root)
+	if !ok {
+		// Nothing found in db. Exit early.
+		return nil, nil
+	}
+
+	// Compute the file path.
+	filePath := filePath(root, summary.epoch)
+
+	// Open the data column sidecars file.
+	// We do not specially check if the file exists since we have already checked the cache.
+	file, err := dcs.fs.Open(filePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "data column sidecars file path open")
+	}
+
+	// Read file metadata.
+	metadata, err := dcs.metadata(file)
+	if err != nil {
+		return nil, errors.Wrap(err, "metadata")
+	}
+
+	// Retrieve data column sidecars from the file.
+	verifiedRODataColumnSidecars := make([]blocks.VerifiedRODataColumn, 0, len(indices))
+	for _, index := range indices {
+		// Skip if the data column is not saved.
+		if metadata.indices[index] < limit {
+			continue
+		}
+
+		// Compute the offset of the data column sidecar.
+		offset := headerSize + int64(metadata.indices[index]-limit)*int64(metadata.sszEncodedDataColumnSidecarSize)
+
+		// Seek to the beginning of the data column sidecar.
+		_, err := file.Seek(offset, io.SeekStart)
+		if err != nil {
+			return nil, errors.Wrap(err, "seek")
+		}
+
+		// Read the SSZ encoded data column sidecar.
+		sszEncodedDataColumnSidecar := make([]byte, metadata.sszEncodedDataColumnSidecarSize)
+		count, err := file.Read(sszEncodedDataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "read SSZ encoded data column sidecar")
+		}
+		if uint32(count) != metadata.sszEncodedDataColumnSidecarSize {
+			return nil, ErrWrongBytesWritten
+		}
+
+		// Unmarshal the SSZ encoded data column sidecar.
+		dataColumnSidecar := &ethpb.DataColumnSidecar{}
+		err = dataColumnSidecar.UnmarshalSSZ(sszEncodedDataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "unmarshal SSZ encoded data column sidecar")
+		}
+
+		// Create a RO data column.
+		roDataColumnSidecar, err := blocks.NewRODataColumn(dataColumnSidecar)
+		if err != nil {
+			return nil, errors.Wrap(err, "new read only data column")
+		}
+
+		// Create a verified RO data column.
+		verifiedRODataColumn := blocks.NewVerifiedRODataColumn(roDataColumnSidecar)
```

Comment:
Since this section creates a `VerifiedRO` type I would prefer for it to be in the `verification` package. I realize this leaks some storage details into the other package, which feels wrong, but I want to be strict about where we call the `NewVerifiedRO` types so we can enforce it with a static analyzer. I'm making a note to self to write that analyzer now. I think [ReadFull](https://pkg.go.dev/io#ReadFull) could be helpful again here - you give the buffer and the `Reader` to the func in `verification` and it returns the verified ro type.

Reply:
Awesome!